src/requirements/reqs-functions.ts

```diff
@@ -171,7 +158,74 @@ function computeFulfillmentStatistics<T extends {}>({ requirement, courses: cour
     }
   });
 
-  return { minCountFulfilled, totalCountFulfilled };
+  return {
+    minCountFulfilled,
+    minCountRequired,
+    totalCountFulfilled,
+    totalCountRequired,
+    courses: coursesThatFulfilledRequirement
+  };
+}
+
+/**
+ * @param coursesTaken a list of all taken courses.
+ * @param requirementCourses a list of eligible courses from requirement data.
+ * @returns a naively computed list of courses that fulfill the requirement, partitioned into sub-requirement filfillment.
+ */
+function filterAndPartitionCoursesThatFulfillRequirement(
+  coursesTaken: readonly CourseTaken[],
+  requirementCourses: readonly EligibleCourses[],
+): CourseTaken[][] {
+  const coursesThatFulfilledRequirement: CourseTaken[][] = requirementCourses.map(() => []);
+  coursesTaken.forEach(courseTaken => {
+    const { roster, courseId } = courseTaken;
+    requirementCourses.forEach((subRequirementCourses, subRequirementIndex) => {
+      if (subRequirementCourses[roster] && subRequirementCourses[roster].includes(courseId)) {
+        // add the course to the list of courses used to fulfill that one sub-requirement
+        coursesThatFulfilledRequirement[subRequirementIndex].push(courseTaken);
+      }
+    });
+  });
+  return coursesThatFulfilledRequirement;
+}
+
+function computeFulfillmentCoursesAndStatistics(
+  requirement: DecoratedCollegeOrMajorRequirement,
+  coursesTaken: readonly CourseTaken[],
+): RequirementFulfillmentStatistics & { readonly courses: readonly (readonly CourseTaken[])[] } {
+  switch (requirement.fulfilledBy) {
+    case 'self-check':
+      return { minCountFulfilled: 0, minCountRequired: 0, courses: [] };
+    case 'courses':
+    case 'credits':
+      return computeFulfillmentStatisticsFromCourses(
+        filterAndPartitionCoursesThatFulfillRequirement(coursesTaken, requirement.courses),
+        requirement.fulfilledBy,
+        requirement.operator,
+        requirement.minCount,
+        requirement.totalCount
+      );
+    case 'toggleable': {
+      // Choose the max fulfillment progress as the statistics
+      // TODO: take user choices into account when we are going to integrating that.
```

Comment:
For this specific part, are you planning on just copying over a lot of Ein's dropdown code?

Reply:
This will be the eventual goal, but in the next PR, I'm just doing to add the toggleable requirement without doing this. Integrating Ein's change also implies it needs to be integrated with the graph, which can be more involved.